Homework for sprint №17 (DI)

sobolevkir opened the pull request.

sobolevkir committed:
Использовано внедрение зависимостей через библиотеку Koin

ArturNurtdinov commented on app/src/main/java/com/sobolevkir/playlistmaker/di/InteractorModule.kt:
```diff
@@ -0,0 +1,21 @@
+import org.koin.dsl.module
+
+val interactorModule = module {
+    single<FavoritesInteractor> { FavoritesInteractorImpl(get()) }
+    single<TracksInteractor> { TracksInteractorImpl(get(), get()) }
+    factory<PlayerInteractor> { PlayerInteractorImpl(get()) }
+    single<SettingsInteractor> { SettingsInteractorImpl(get()) }
+    single<SharingInteractor> { SharingInteractorImpl(get(), get()) }
```
Интеракторы в общем случае не являются тяжёлыми объектами и не хранят состояние, их очень легко создать. Поэтому их не принято делать синглами, а при необходимости создавать новый объект, чтобы не тратить лишнюю память на их хранение, поэтому здесь лучше сделать через `factory`. С репозиториями - обратная ситуация, они обычно хранят состояние и являются тяжёлыми объектами, их лучше делать через `single`

sobolevkir replied:
А репозиторий плеера если сделать sigle, то плеер не будет запускаться при закрытии активити одного трека и открытии другого. Как в этом случае поступить?

sobolevkir replied:
мы же используем метод release при закрытии активити плеера. и после выполнения этого метода уже невозможно использовать плеер повторно. а т.к. в конструкторе репозитория передаётся экземпляр MediaPlayer, то он уже будет непригоден для использования, не так ли?

sobolevkir replied:
не совсем понял. Ну вот создаём мы медиаплеер внутри репозитория. Но ведь этот объект будет один и тот же и опять же после release мы уже не сможем с ним работать?

ArturNurtdinov commented on app/src/main/java/com/sobolevkir/playlistmaker/player/ui/activity/PlayerActivity.kt:
```diff
@@ -13,15 +12,18 @@ import com.sobolevkir.playlistmaker.common.domain.model.Track
-        PlayerViewModel.getViewModelFactory(track ?: Track())
+    private val viewModel: PlayerViewModel by viewModel {
+        parametersOf(
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                intent.getSerializableExtra(CURRENT_TRACK, Track::class.java)
+            } else intent.getSerializableExtra(CURRENT_TRACK) as Track
```
Это можно в отдельную функцию вынести, а лучше в экстеншн функцию к классу Intent, такие функции в андроиде принято называть с припиской Compat, это говорит о том, что функция поддерживает больший набор версий андроид, чем оригинальная функция. Например функция `Intent.getSerializableExtraCompat(key: String, clazz: Class<*>?)` будет поддерживать так же работу на апи ниже TIRAMISU

sobolevkir replied:
принято 🫡

sobolevkir replied:
В общем, с таким методом приложение падает:
<img width="580" alt="image" src="https://github.com/sobolevkir/PlaylistMaker/assets/108625077/1fd9a39a-3309-456a-ad60-be1909433375">
А вот с таким приложение не компилируется:
<img width="579" alt="image" src="https://github.com/sobolevkir/PlaylistMaker/assets/108625077/64d982e2-e910-4d3f-aea3-22a3ba7e5aa9">
Как тогда лучше поступить?

sobolevkir committed:
Небольшие корректировки:
1. Проверка версии при получении Serializable через Intent вынесена в функцию-расширение Intent.getSerializableExtraCompat.
2. Player и экземпляр MediaPlayer теперь помечены в DI как Single.
